### PR TITLE
Turn the screen off while using ear speaker

### LINKED ISSUE #1735: Call screen does not use proximity sensor
In a voice call holding the phone to the ear does not turn off the display so touching it with your ear or cheek is opening up the notification drawer and doing some other unwanted things.

### PR DESCRIPTION
**Pull request by onurays**
Fixes #1735

### CODE REVIEW

File: vector/src/main/java/im/vector/app/core/services/VectorService.kt [hunk 1]
```diff
@@ -33,10 +31,6 @@ abstract class VectorService : Service() {
      */
     private var mIsSelfDestroyed = false
 
-    override fun attachBaseContext(base: Context) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/core/services/VectorService.kt:36]: Why is this removed?

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * Start listening the proximity sensor. [stop] function should be called to release the sensor and the WakeLock.
+     */
+    fun start() {
+        powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
+        sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
+        sensor = sensorManager.getDefaultSensor(Sensor.TYPE_PROXIMITY)
+
+        if (sensor != null && powerManager.isWakeLockLevelSupported(PowerManager.PROXIMITY_SCREEN_OFF_WAKE_LOCK)) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:50]: could it be really null?

Maybe we need some synchronisation?

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_NORMAL)
+        }
+    }
+
+    /**
+     * Stop listening proximity sensor changes and release the WakeLock.
+     */
+    fun stop() {
+        if (this::sensorManager.isInitialized) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:59]: why not initialize in constructor if it's lazy?

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        }
+        wakeLock?.release()
+    }
+
+    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
+        // NOOP
+    }
+
+    override fun onSensorChanged(event: SensorEvent) {
+        val distanceInCentimeters = event.values[0]
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:70]: In the app rtc reference app, they doing a check against sensor type, maybe we should do it too? if (event.sensor.getType() != Sensor.TYPE_PROXIMITY) return;

File: vector/src/main/java/im/vector/app/features/call/VectorCallViewModel.kt [hunk 5]
```diff
@@ -145,10 +146,17 @@ class VectorCallViewModel @AssistedInject constructor(
         }
 
         override fun onAudioDevicesChange(mgr: WebRtcPeerConnectionManager) {
+            val currentSoundDevice = mgr.audioManager.getCurrentSoundDevice()
+            if (currentSoundDevice == CallAudioManager.SoundDevice.PHONE) {
```
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/call/VectorCallViewModel.kt:150]: Why only do it if on ear peace?

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]
Please see my comments

File: vector/src/main/java/im/vector/app/core/services/VectorService.kt (hunk 1 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/core/services/VectorService.kt:36]: It is crashing the application. @bmarty suggested to remove it.

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt (hunk 2 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:50]: Yes, the official documentation says some devices may not have the proximity sensor.
https://developer.android.com/guide/topics/sensors/sensors_position#sensors-pos-prox

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt (hunk 3 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt]: It is not lazy, but there is a change that `stop()` may be called before `start()`. But it is ugly, I moved initializations to the constructor.

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt (hunk 4 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:70]: We are pretty sure we have only proximity sensor here since we don't register any other sensors. I think it is so defensive coding.

File: vector/src/main/java/im/vector/app/features/call/VectorCallViewModel.kt (hunk 5 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/call/VectorCallViewModel.kt:150]: We are trying to protect the screen against undesired touches with the ear. Do you think we have another use case?

File: vector/src/main/java/im/vector/app/core/services/VectorService.kt (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/core/services/VectorService.kt:36]: @bmarty any thought?

File: vector/src/main/java/im/vector/app/features/call/VectorCallViewModel.kt (hunk 5 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/call/VectorCallViewModel.kt:150]: I think there could be some use to switch from earpiece/speaker depending on proximity, but maybe as an option

File: vector/src/main/java/im/vector/app/core/services/VectorService.kt (hunk 1 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/core/services/VectorService.kt:36]: This is OK to remove it, I think this is not really necessary for Services. It was added recently (after the previous release) to have a localized Context. It's probably useless for Services.

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    private lateinit var powerManager: PowerManager
+    private lateinit var sensorManager: SensorManager
+
+    private var wakeLock: PowerManager.WakeLock? = null
+    private var sensor: Sensor? = null
+
+    init {
+        powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
+        sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:44]: Please use `getSystemService<Class>` (see https://github.com/vector-im/element-android/pull/1702)

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            onProximityNear()
+        } else {
+            onProximityFar()
+        }
+    }
+
+    /**
+     * Recommending naming convention for WakeLock tags is "app:tag"
+     */
+    private fun generateWakeLockTag() = "${context.getString(R.string.app_name)}:$PROXIMITY_WAKE_LOCK_TAG"
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt:81]: You should inject and use the StringProvider to get Strings

**Review by bmarty — CHANGES_REQUESTED** [review]
2 remarks for code quality

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt (hunk 6 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt]: Done.

File: vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt (hunk 7 above)
Comment (onurays) [reply vector/src/main/java/im/vector/app/features/call/CallProximityManager.kt]: Done.

**Review by bmarty — APPROVED** [review]
